src/translator.tsx

```diff
@@ -0,0 +1,76 @@
+import * as React from 'react';
+
+export type TranslationHandler = {
+  (value: string, setTranslation: (translation: string) => void): Promise<void>;
+};
+
+const defaultHandler: TranslationHandler = () => new Promise(() => {});
+
+export const TranslateContext: React.Context<
+  TranslationHandler
+> = React.createContext(defaultHandler);
+export const LanguageContext: React.Context<string> = React.createContext('en');
+
+type CacheProvider = {
+  get: (language: string, key: string) => string | undefined;
+  set: (language: string, key: string, translation: string) => void;
+};
+
+type Props = {
+  to: string;
+  from: string;
+  cacheProvider?: CacheProvider;
+  children: string;
+  googleApiKey: string;
+};
+
+export default function Translator({
+  to,
+  from,
+  cacheProvider,
+  children,
+  googleApiKey,
+}: Props): JSX.Element {
+  const handleTranslationAsync: TranslationHandler = async (
+    value,
+    setTranslation
+  ) => {
+    let useCache = !!cacheProvider;
+    let translatedText = value;
+
+    if (to !== from && cacheProvider) {
+      const cachedTranslation = cacheProvider.get(to, value);
+
+      if (cachedTranslation) {
+        translatedText = cachedTranslation;
```

Comment:
Could we set our translation and return here as well? Would save the `if (!useCache)` below, right - since we know we'd need it?